### PR TITLE
Add new option --with-sanitize[=list,of,clang,sanitizers]

### LINKED ISSUE #698: [0.4.1] The gem is not correctly compiled with GCC 4.8.4
The installation of the gem is fine, however when it's put to use by some DB task, the following error is shown.

```bash
rake aborted!
LoadError: ./vendor/bundle/ruby/2.1.0/gems/mysql2-0.4.1/lib/mysql2/mysql2.so: undefined symbol: __asan_unregister_globals - /home/runner/productive/vendor/bundle/ruby/2.1.0/gems/mysql2-0.4.1/lib/mysql2/mysql2.so
```

## Reproducing the error
```bash
~ $ gcc --version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4
# clone a sample Rails 3 project
~ $ git clone https://github.com/thinreports/thinreports-rails3-example.git && cd thinreports-rails3-example
# add the mysql2 (0.4.1) gem to the Gemfile
# create ./config/database.yml and configure it to use the mysql2 gem
# install the gems
thinreports-rails3-example $ bundle install --path vendor/bundle
# run any DB task with rake
thinreports-rails3-example $ bundle exec rake db:drop
# !!! the error is shown here
```

## Mitigating the issue
Older versions of the gem don't have this issue. To mitigate it, an older GCC has to be used.

```
$ sudo apt-get install -y gcc-4.7 g++-4.7
$ sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.7 100 --slave /usr/bin/g++ g++ /usr/bin/g++-4.7
$ sudo update-alternatives --auto gcc
$ gcc --version
gcc (Ubuntu/Linaro 4.7.3-12ubuntu1) 4.7.3
# remove the installed gems from the sample project,
# of course removing only the mysql2 gem works too
thinreports-rails3-example $ rm -rf vendor
# install the gems again
thinreports-rails3-example $ bundle install --path vendor/bundle
# run a DB task with rake
thinreports-rails3-example $ bundle exec rake db:drop
# the command shouldn't error out
```

The OS is Ubuntu 14.04 x64.

### PR DESCRIPTION
**Pull request by sodabrew**
Resolves #698, #695, #685.

/cc @tamird

### CODE REVIEW

File: ext/mysql2/extconf.rb [hunk 1]
```diff
@@ -105,31 +105,52 @@ def asplode(lib)
...
 $CFLAGS << ' ' << usable_flags.join(' ')
 
+enabled_sanitizers = disabled_sanitizers = []
+# Specify a commna-separated list of sanitizers, or try them all by default
+sanitizers = with_config('sanitize')
+case sanitizers
+when true
+  # Try them all, turn on whatever we can
+  sanitizers = 'address,cfi,integer,memory,thread,undefined'
+  enabled_sanitizers, _ = sanitizers.split(',').partition do |s|
```
Comment (tamird) [inline ext/mysql2/extconf.rb]: s/partition/select/

**Comment by tamird** [comment]
LGTM. Do you want these sanitizers enabled in CI?

File: ext/mysql2/extconf.rb (hunk 1 above)
Comment (sodabrew) [inline ext/mysql2/extconf.rb]: Yep since I'm ignoring the false array anyhow.

**Comment by sodabrew** [comment]
At the moment, only the Travis OS X instances are compiling with new enough clang or GCC, and for some reason I haven't figured out yet, mysql2 compiled with address sanitizer hangs on OS X. So for now I'll leave these off in CI. When Travis moves to Ubuntu Trusty, its gcc does include address sanitizer, and I'll revisit.

**Pull request merged**